lib/sycamore/sycamore/document.py

```diff
@@ -0,0 +1,388 @@
+from abc import abstractmethod, ABC
+from enum import Enum
+
+
+class Category(Enum):
+    Caption = 1
+    Footnote = 2
+    Formula = 3
+    ListItem = 4
+    PageFooter = 5
+    PageHeader = 6
+    Picture = 7
+    SectionHeader = 8
+    Table = 9
+    Text = 10
+    Title = 11
+    Section = 101
+    Group = 102
+
+
+class Node:
+    def __init__(self, node_id, metadata):
+        self._node_id = node_id
+        self._metadata = metadata
+
+    def node_id(self):
+        return self._node_id
+
+    @abstractmethod
+    def category(self):
+        pass
+
+    def metadata(self):
+        return self._metadata
+
+    @abstractmethod
+    def accept(self, visitor):
+        pass
+
+
+class Leaf(Node):
+    def __init__(self, node_id, box, content, metadata):
+        super().__init__(node_id, metadata)
+        self._box = box
+        self._content = content
+
+    def box(self):
+        return self._box
+
+    def content(self):
+        return self._content
+
+
+class Internal(Node):
+    def __init__(self, node_id,  children, metadata):
+        super().__init__(node_id, metadata)
+        self._children = children
+
+    def children(self) -> list[Node]:
+        return self._children
+
+
+class Caption(Leaf):
```

Comment:
Certainly not critical, but would it make sense to create a parent class for all of the text-like leaf nodes (e.g. list items, text, etc.) I could imagine a lot of operations where you want to do something special for tables and images, but something common for all the text.

Reply:
yes, sure.